### PR TITLE
Refactor of the dynamic provider project to have the core much smalle…

### PR DESCRIPTION
**Pull request by csteeg**
Ok, so this one is quite large, but I hope you like it.

This is what I needed to do to be able to pull in your project into our solution.
It was not possible for us to implement your EF-Core nuget package, since it required us to inherit from your context & baseclass and we don't want to add a nuget package to our shared project that contains our simple objects (POCO)... it just adds too much dependencies which we don't need and want. 

This refactor, makes that the core has nothing to do with storing data anymore, it simply doesn't care where the data is coming from or how it looks, as long as it has a store that can return a enumerable of ISchemeDefinition

Our implementation now just has to implement ISchemeDefinition in a simple dummy-class and that dummy class is returned by our own IDynamicProviderStore which only does some very simple conversion from our POCO/storage implemenation to this interfaced object

### CODE REVIEW

**Comment by aguacongas** [comment]
@csteeg Thanks to contribute. You did a great job, however, I'm not sure your solution works if servers are load balanced and you need to refresh the configuration because you removed `PersistentDynamicManager` and `NoPersistentDynamicManager` from the DI. There is no way to notify the configuration changes and reload it.  
Did you try to create a POC with a 2 instances of the server, change the configuration on one side and check if it's updated on the other side ?

**Comment by csteeg** [comment]
@aguacongas yes, our system runs on multiple pods also. I use rabbitmq for messaging, but that doesn't matter ofcourse :)... Your admin-project uses SignalR IIRC

On our project we trigger an event when any dynamicprovider change happens (beit a delete, update or add). In the eventhandler on all pods we just do:
```
            IEnumerable<ISchemeDefinition> items = await dynamicProviderStore.GetSchemeDefinitionsAsync(cancellationToken).ToListAsync().ConfigureAwait(false);
            logger.LogInformation($"Received message that scheme {message.SchemeName} was changed");

            await authenticationSchemeProviderWrapper.InitializeAsync(items, cancellationToken).ConfigureAwait(false);

```
The InitializeAsync does updates, adds, deletes based on the new items from the dynamicproviderstore.

The event triggering wasn't part of the project before this refactor also right?
I think it would be best to do so actually, I'll add it to your project. The default could then be a null-implementation and people could add it the way the want (like with signalR in your case, rabbitmq in my case)

**Comment by aguacongas** [comment]
@csteeg I agree. Also if you can update the sample and readme files as well it'll be great

**Comment by aguacongas** [comment]
@csteeg Ho, and add your name in licenses headers of files you changes and add if you want.

**Comment by csteeg** [comment]
@aguacongas will update in a few days time.. bit busy these coming days.. should be friday or saturday

**Comment by csteeg** [comment]
@aguacongas I've added a base eventhandler which should be easy to override on a multiple-container implementation and updated the sample (though had to point the sample to the project files instead of the packages because of the breaking changes)

File: sample/Aguacongas.AspNetCore.Authentication.Sample/Controllers/HomeController.cs [hunk 1]
```diff
@@ -133,12 +141,12 @@ public async Task<IActionResult> Update(AuthenticationViewModel model)
         [Route("List")]
         public async Task<IActionResult> List()
         {
-            var schemes = await _signInManager.GetExternalAuthenticationSchemesAsync();
+            System.Collections.Generic.IEnumerable<AuthenticationScheme> schemes = await _signInManager.GetExternalAuthenticationSchemesAsync();
```
Comment (aguacongas) [inline sample/Aguacongas.AspNetCore.Authentication.Sample/Controllers/HomeController.cs:144]: can you simplify by adding a using System.Collections.Generic or using var please 
As you probably note. I prefer using var, it's easier to refactor.

File: sample/Aguacongas.AspNetCore.Authentication.Sample/Controllers/HomeController.cs [hunk 2]
```diff
@@ -133,12 +141,12 @@ public async Task<IActionResult> Update(AuthenticationViewModel model)
...
 
-            var managedSchemes = schemes.Where(s => _manager.ManagedHandlerType.Any(h => s.HandlerType == h))
+            System.Collections.Generic.IEnumerable<string> managedSchemes = schemes.Where(s => _handerTypeProvider.GetManagedHandlerTypes().Any(h => s.HandlerType == h))
```
Comment (aguacongas) [inline sample/Aguacongas.AspNetCore.Authentication.Sample/Controllers/HomeController.cs:146]: can you simplify by adding a using System.Collections.Generic or using var please

File: sample/Aguacongas.AspNetCore.Authentication.Sample/Controllers/HomeController.cs [hunk 3]
```diff
@@ -133,12 +141,12 @@ public async Task<IActionResult> Update(AuthenticationViewModel model)
...
                 .Select(s => s.Name);
 
-            var definitions = managedSchemes.Select(name => _manager.FindBySchemeAsync(name).GetAwaiter().GetResult());
+            System.Collections.Generic.IEnumerable<SchemeDefinition> definitions = managedSchemes.Select(name => _mutationStore.FindBySchemeAsync(name).GetAwaiter().GetResult());
```
Comment (aguacongas) [inline sample/Aguacongas.AspNetCore.Authentication.Sample/Controllers/HomeController.cs:149]: can you simplify by adding a using System.Collections.Generic or using var please

**Review by aguacongas — CHANGES_REQUESTED** [review]
Please use var keyword everywhere it's possible. You re typed everythings but that do to much changes to reviewed. And I prefer using var when it's possible because it's easier to refactor.

**Comment by csteeg** [comment]
Sorry about that, forgot to turn off my code-cleanup plugin with my own ruleset

File: src/Aguacongas.AspNetCore.Authentication.TestBase/DynamicManagerTestBase.cs [hunk 4]
```diff
@@ -34,7 +34,7 @@ namespace Aguacongas.AspNetCore.Authentication.TestBase
     /// </summary>
     /// <typeparam name="TSchemeDefinition">The type of the scheme definition.</typeparam>
     public abstract class DynamicManagerTestBase<TSchemeDefinition>
```
Comment (aguacongas) [inline src/Aguacongas.AspNetCore.Authentication.TestBase/DynamicManagerTestBase.cs:36]: The class should be renamed `AuthenticationSchemeProviderWrapperTestBase` IMO as `DynamicManager` doesn't exist any more

**Review by aguacongas — CHANGES_REQUESTED** [review]
Great job. 
Can you document the breaking change in a Breaking_Change.md or in the README.md? All we need to do to upgrade from the previous version to this one, please.

**Pull request closed without merging**